File: Data Structures & Algorithms/reverse-a-linked-list/submission-4.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {

        if(head == null) {
            return head;
        }

        ArrayList<Integer> array = new ArrayList<>();
        while(head != null) {
            array.add(head.val);
            head = head.next;
        }
        Collections.reverse(array);

        ListNode newHead = new ListNode(array.get(0));
        ListNode current = newHead;

        for(int i = 1; i < array.size(); i++) {
            current.next = new ListNode(array.get(i));
            current = current.next;
        }

        return newHead;
    }
        
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        current, array = head, list()
        while current:
            array.append(current.val)
            current = current.next
        
        def split(array: list, k: int):
            for i in range(0, len(array), k):
                yield array[i:i+k]
        
        split = list(split(array, k))

        for i in range(len(split)):
            if len(split[i]) == k:
                split[i].reverse()
        
        current = head
        newHead = current
        for i in range(len(split)):
            for j in range(len(split[i])):
                current.next = ListNode(split[i][j])
                current = current.next
        
        return newHead.next
        